src/NaluCommNeighbors.C

```diff
@@ -0,0 +1,156 @@
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+#include <stdlib.h>
+#include <stdexcept>
+#include <sstream>
+#include <iostream>
+#include <vector>
+#include <algorithm>
+
+#include <stk_util/util/SortAndUnique.hpp>
+#include <stk_util/environment/CPUTime.hpp>
+#include <stk_util/parallel/ParallelComm.hpp>
+#include <NaluCommNeighbors.hpp>
+
+namespace nalu_stk {
+
+//-----------------------------------------------------------------------
+
+#if defined( STK_HAS_MPI )
+
+void CommNeighbors::rank_error( const char * method , int p ) const
+{
+  std::ostringstream os ;
+  os << "stk::CommNeighbors::" << method
+     << "(" << p << ") ERROR: Not in [0:" << m_size << ")" ;
+  throw std::range_error( os.str() );
+}
+
+//----------------------------------------------------------------------
+
+CommNeighbors::~CommNeighbors()
+{
+  m_comm = stk::parallel_machine_null();
+  m_size = 0 ;
+  m_rank = 0 ;
+}
+
+CommNeighbors::CommNeighbors( stk::ParallelMachine comm, const std::vector<int>& neighbor_procs)
+  : m_comm( comm ),
+    m_size( stk::parallel_machine_size( comm ) ),
+    m_rank( stk::parallel_machine_rank( comm ) ),
+    m_send(),
+    m_recv(),
+    m_send_data(),
+    m_recv_data(),
+    m_send_procs(neighbor_procs),
+    m_recv_procs(neighbor_procs)
+{
+  m_send.resize(m_size);
+  m_recv.resize(m_size);
+  std::sort(m_neighbor_procs.begin(), m_neighbor_procs.end());
+}
+
+CommNeighbors::CommNeighbors( stk::ParallelMachine comm, const std::vector<int>& send_procs, const std::vector<int>& recv_procs)
+  : m_comm( comm ),
+    m_size( stk::parallel_machine_size( comm ) ),
+    m_rank( stk::parallel_machine_rank( comm ) ),
+    m_send(),
+    m_recv(),
+    m_send_data(),
+    m_recv_data(),
+    m_send_procs(send_procs),
+    m_recv_procs(recv_procs)
+{
+  m_send.resize(m_size);
+  m_recv.resize(m_size);
+  stk::util::sort_and_unique(m_send_procs);
+  stk::util::sort_and_unique(m_recv_procs);
+}
+
+void CommNeighbors::communicate()
+{
+  const int mpitag = 10101, mpitag2 = 10102;
+  int maxRecvProcs = m_recv_procs.size();
+  int maxSendProcs = m_send_procs.size();
+  int max = std::max(maxSendProcs, maxRecvProcs);
+  std::vector<int> recv_sizes(maxRecvProcs, 0);
+  std::vector<MPI_Request> requests(max, MPI_REQUEST_NULL);
+  std::vector<MPI_Request> requests2(max, MPI_REQUEST_NULL);
+  std::vector<MPI_Request> requests3(max, MPI_REQUEST_NULL);
+  std::vector<MPI_Status> statuses(max);
+  for(int i=0; i<maxRecvProcs; ++i) {
```

Comment:
My suspicion was correct: this is another boundary exchange reimplementation.

Use Tpetra::Distributor, or Dan Ibanez's interface, or MPI 3's, or maybe STK's (because surely it has one).  Don't reimplement this.  Boundary exchange implementations may need to change a LOT in the next few years for good performance.  The fewer reimplementations we have to rewrite, the better.

Reply:
@mhoemmen you must be referring to MPI_Neighbor_alltoallv. I'll investigate that, thanks.

Reply:
Tpetra's doesn't use a global sync.  Neither does the MPI 3 version, and I doubt that Dan's does either.

See Chapter 7 of the MPI 3.0 Standard for neighborhood collectives and process topologies.  You would use `MPI_Dist_graph_create_adjacent` to create the communication pattern (see Example 7.3, pp. 300-1, esp. the last table in the example).  The pattern gets attached to a new `MPI_Comm` which is otherwise a clone of the input `MPI_Comm`.  You then pass the resulting `MPI_Comm` to either of the following:

  - `MPI_Neighbor_alltoall` (see Section 7.6.2, p. 319), if you need to send a fixed number of things to/from each process, or
  - `MPI_Neighbor_alltoallv` (p. 321), if you need to send a variable number of things to/from each process (which should be the case here)

The corresponding `Tpetra::Distributor` pattern would be: 
  1. Create Distributor using `createFromSendsAndRecvs`
  2. Use either three-argument doPostsAndWaits (for the "alltoall" case above), or four-argument doPostsAndWaits (for the "alltoallv" case above)

Reply:
@mhoemmen thanks. I'll confess I didn't know the details of Tpetra::Distributor, and have never heard of Dan's. The MPI one wasn't available to us in sierra until we upgraded to open-mpi 1.10. (Which is still not new, but at least it has the MPI_Neighbor_* functions.)

Reply:
@alanw0 Dan's isn't in Trilinos (yet! ;-) ).

`Tpetra::Distributor` (and `Epetra_*Distributor`) also implement boundary exchange.  They are the communication layer that doExport and doImport use underneath.

Reply:
@mhoemmen  I've pushed another commit, addressing your comments (removing ArrayRCP in favor of View, and using MPI_Neighbor boundary exchange).